langchain4j-vertex-ai-gemini/src/test/java/dev/langchain4j/model/vertexai/VertexAiGeminiChatModelIT.java

```diff
@@ -456,28 +491,186 @@ void should_use_tools_with_AiService_2() {
+        public String paymentStatus(@P("The ID of the payment transaction") String transactionId) {
+            System.out.println("PAYMENT STATUS FOR TRANSACTION " + transactionId + " CALLED");
+            return "Transaction " + transactionId + " is " + DATASET.get(new Transaction(transactionId));
+        }
+    }
+
+    interface FunctionCallingAssistant {
+        @dev.langchain4j.service.SystemMessage("You MUST call the `paymentStatus()` function if you're asked about a payment transaction.")
+        String chat(String userMessage);
+    }
+
+    @Test
+    void should_work_with_interspersed_function_execution_results() {
+        // given
+        VertexAiGeminiChatModel model = VertexAiGeminiChatModel.builder()
+            .project("genai-java-demos")
+            .location("us-central1")
+            .modelName("gemini-1.5-flash-001")
+            .temperature(0.0f)
+            .topK(1)
+            .logRequests(true)
+            .logResponses(true)
+            .maxRetries(1)
+            .build();
+
+        FunctionCallingService service = new FunctionCallingService();
+
+        FunctionCallingAssistant assistant = AiServices.builder(FunctionCallingAssistant.class)
+            .chatLanguageModel(model)
+            .chatMemory(MessageWindowChatMemory.withMaxMessages(100))
+            .tools(service)
+            .build();
+
+        // when
+        String response = assistant.chat("What is the status of my payment transactions 001 and 002?");
+        System.out.println(response);
+
+        // then
+        assertThat(response).contains("001");
+        assertThat(response).contains("002");
+        assertThat(response).contains("pending");
+        assertThat(response).contains("approved");
+        assertThat(response).doesNotContain("003");
+        assertThat(response).doesNotContain("rejected");
+
+        // when
+        response = assistant.chat("What is the status of transactions 003?");
+        System.out.println(response);
+
+        // then
+        assertThat(response).doesNotContain("001");
+        assertThat(response).doesNotContain("002");
+        assertThat(response).doesNotContain("pending");
+        assertThat(response).doesNotContain("approved");
+        assertThat(response).contains("003");
+        assertThat(response).contains("rejected");
+    }
+
+    @Test
+    void should_use_google_search() {
+
+        // given
+        VertexAiGeminiChatModel modelWithSearch = VertexAiGeminiChatModel.builder()
+            .project("genai-java-demos")
+            .location("us-central1")
+            .modelName("gemini-1.5-flash-001")
+            .useGoogleSearch(true)
+            .build();
+
+        // when
+        String resp = modelWithSearch.generate(
+            "Why is the sky blue?");
+
+        // then
+        System.out.println("Google Search powered response = " + resp);
+        assertThat(resp).contains("scatter");
```

Comment:
hmm, does it prove that google search was used?

Reply:
Not really. That's right.
At first, I had a test that worked very well, as it was searching for a very recent news item.
But after a few days, the news item disappeared from the Google news headlines, and the search wouldn't give me the result I was expecting.
So this is very hard to test, and not make the test flaky (and have to change it every week)